assignment/tool/src/java/org/sakaiproject/assignment/entityproviders/AssignmentEntityProvider.java

```diff
@@ -457,6 +460,261 @@ public GraderUser(User sakaiUser) {
+            		+ "\"code\": 1,"
+            		+ "\"message\": \"ts.add.err.regTime\""
+            		+ "}"
+            		+ "}";
+        }
+
+        String asnComment = (String) params.get("regComment");
+        StringBuilder alertMsg = new StringBuilder();
+        asnComment = formattedText.processFormattedText(asnComment, alertMsg);
+        if (alertMsg.length() > 0) {
+            log.warn("Comment field format is not valid");
+            return "{\"error\": {"
+                    + "\"code\": 1,"
+                    + "\"message\": \"ts.add.err.regComment\""
+                    + "}"
+                    + "}";
+        }
+        
+        Instant regDate = Instant.now();
+        try {
+            int month = Integer.parseInt((String)params.get("new_ts_record_month"));
+            int day = Integer.parseInt((String)params.get("new_ts_record_day"));
+            int year = Integer.parseInt((String)params.get("new_ts_record_year"));
+            int hour = Integer.parseInt((String)params.get("new_ts_record_hour"));
+            int min = Integer.parseInt((String)params.get("new_ts_record_minute"));
+            
+            regDate = LocalDateTime.of(year, month, day, hour, min, 0).atZone(userTimeService.getLocalTimeZone().toZoneId()).toInstant();
+        }catch (NumberFormatException nfe) {
+        	regDate = Instant.now();
+        }
+
+        AssignmentSubmissionSubmitter ass = as.getSubmitters().stream().findAny().orElse(null);
+
+        if(ass == null) {
+        	log.warn("You submitter does not exist");
+            return "{\"error\": {"
+            		+ "\"code\": 1,"
+            		+ "\"message\": \"ts.add.err.submitter\""
+            		+ "}"
+            		+ "}";
+        }
+
+        String context = ass.getSubmission().getAssignment().getContext();
+        
+        AssignmentTimeSheet timeSheet = new AssignmentTimeSheet();
+        timeSheet.setAsnComment(asnComment);
+        timeSheet.setRegDate(regDate);
+        timeSheet.setRegTime(regTime);
+        timeSheet.setSubmitter(ass);
+        
+        try {
+			assignmentService.addAssignmentTimeSheet(timeSheet, context);
+		} catch (PermissionException e) {
+            log.warn("You can't modify this sumbitter");
+            return "{\"error\": {"
+            		+ "\"code\": 1,"
+            		+ "\"message\": \"ts.add.err.permission\""
+            		+ "}"
+            		+ "}";
+		}
+
+        return "{\"success\": true}";
+    }
+
+    @EntityCustomAction(action = "removeTimeSheet", viewKey = EntityView.VIEW_NEW)
+    public String removeTimeSheet(Map<String, Object> params) {
+
+        String userId = sessionManager.getCurrentSessionUserId();
+
+        if (StringUtils.isBlank(userId)) {
+            log.warn("You need to be logged in to add time sheet register");
+            return "{\"error\": {"
+            		+ "\"code\": 1,"
+            		+ "\"message\": \"ts.rem.err.userId\""
+            		+ "}"
+            		+ "}";
```

Comment:
You should really return an object here and let EB do the serialization for you. Hand crafting JSON like this is really error prone.

Reply:
done.

Reply:
@ern 
Adrian is right, that is why we created the `BuildTimeSheetReturnMessage` class to mold the return messages around the EntityProvider.
Although it is more difficult to read, it prevents errors from being made when constructing messages.